paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/SortCompactProcedure.java

```diff
@@ -0,0 +1,164 @@
+            new StructType(
+                    new StructField[] {
+                        new StructField("result", DataTypes.BooleanType, true, Metadata.empty())
+                    });
+
+    protected SortCompactProcedure(TableCatalog tableCatalog) {
+        super(tableCatalog);
+    }
+
+    @Override
+    public ProcedureParameter[] parameters() {
+        return PARAMETERS;
+    }
+
+    @Override
+    public StructType outputType() {
+        return OUTPUT_TYPE;
+    }
+
+    @Override
+    public InternalRow[] call(InternalRow args) {
+        Preconditions.checkArgument(args.numFields() >= 3);
+        Identifier tableIdent = toIdentifier(args.getString(0), PARAMETERS[0].name());
+        String sortType = args.getString(1);
+        List<String> sortColumns = Arrays.asList(args.getString(2).split(","));
+
+        String partitionFilter = args.isNullAt(3) ? null : toWhere(args.getString(3));
+
+        return modifyPaimonTable(
+                tableIdent,
+                table -> {
+                    Preconditions.checkArgument(table instanceof FileStoreTable);
+                    InternalRow internalRow =
+                            newInternalRow(
+                                    execute(
+                                            (FileStoreTable) table,
+                                            sortType,
+                                            sortColumns,
+                                            partitionFilter));
+                    return new InternalRow[] {internalRow};
+                });
+    }
+
+    @Override
+    public String description() {
+        return "This procedure execute sort compact action on unaware-bucket table.";
+    }
+
+    private boolean execute(
+            FileStoreTable table,
+            String sortType,
+            List<String> sortColumns,
+            @Nullable String filter) {
+        CoreOptions coreOptions = table.store().options();
+
+        if (!(table instanceof AppendOnlyFileStoreTable) || coreOptions.bucket() != -1) {
+            throw new UnsupportedOperationException(
+                    "Spark sort compact only support unaware-bucket append-only table yet.");
+        }
+
+        Dataset<Row> row = spark().read().format("paimon").load(coreOptions.path().getPath());
+        row = StringUtils.isBlank(filter) ? row : row.where(filter);
+        new WriteIntoPaimonTable(
+                        table,
+                        SaveMode.dynamic(),
+                        TableSorter.getSorter(table, sortType, sortColumns).sort(row),
+                        new Options())
+                .run(spark());
+        return true;
+    }
+
+    @VisibleForTesting
+    static String toWhere(String partitions) {
+        if (StringUtils.isBlank(partitions)) {
+            return null;
+        }
+
+        List<Map<String, String>> maps = ParameterUtils.getPartitions(partitions.split(";"));
+
+        return maps.stream()
```

Comment:
Why use this special format? We can directly pass a where condition string (such as p1 = a and p2 > b) and then let spark analyze it by itself.

Reply:
For compatible with flink action, but iceberg does like this.